Skip loading TypeScript buildinfo when `--reload` flag is passed

Bundled this with `deno bundle --no-check mod.ts` and after that ran `deno run -A --unstable mod.ts` and now this happens. I'm not quite sure how but this happens with a dynamic/worker imports.

Code can be found on [MierenManz/cards](https://github.com/mierenmanz/cards)

```
thread 'worker-0' panicked at 'internal error: entered unreachable code: unexpected missing emit: file:///D:/programming/js_ts/card_game/utils/match/game_errors.ts', 
cli\proc_state.rs:544:15
```

```
stack backtrace:
   0:     0x7ff6174dc3a6 - <unknown>
   1:     0x7ff616c8cc3a - <unknown>
   2:     0x7ff6174da5f8 - <unknown>
   3:     0x7ff6174db462 - <unknown>
   4:     0x7ff6174daa7d - <unknown>
   5:     0x7ff616c419bf - <unknown>
   6:     0x7ff6174da427 - <unknown>
   7:     0x7ff6174e2b35 - <unknown>
   8:     0x7ff6174e2aa9 - <unknown>
  10:     0x7ff6185d5bc0 - CrashForExceptionInNonABICompliantCodeRange
  11:     0x7ff61697b3da - <unknown>
  12:     0x7ff61697888d - <unknown>
  13:     0x7ff616978566 - <unknown>
  14:     0x7ff616fe496e - <unknown>
  15:     0x7ff616fdfc46 - <unknown>
  16:     0x7ff616fdd912 - <unknown>
  17:     0x7ff616fdaa05 - <unknown>
  18:     0x7ff616fd9796 - <unknown>
  19:     0x7ff6174fdd2b - <unknown>
  20:     0x7fffcbcf7034 - BaseThreadInitThunk
  21:     0x7fffcc862651 - RtlUserThreadStart
  ```

Just encountered while updating deps and running a `deno run --allow-net="deno.land" --reload mod.ts` command:

```console
make ci                   
Checked 8 files
Checked 7 files
Download https://deno.land/std@0.115.1/http/server.ts
Download https://deno.land/x/free_port@v1.2.0/mod.ts
Download https://deno.land/x/superdeno@4.7.0/mod.ts
Download https://deno.land/x/superdeno@4.7.0/src/superdeno.ts
Download https://deno.land/x/superdeno@4.7.0/src/test.ts
Download https://deno.land/x/superdeno@4.7.0/version.ts
Download https://deno.land/std@0.115.1/async/mod.ts
Download https://deno.land/x/superdeno@4.7.0/deps.ts
Download https://deno.land/x/superdeno@4.7.0/src/close.ts
Download https://deno.land/x/superdeno@4.7.0/src/types.ts
Download https://deno.land/x/superdeno@4.7.0/src/utils.ts
Download https://deno.land/x/superdeno@4.7.0/src/superagent.ts
Download https://deno.land/x/superdeno@4.7.0/src/xhrSham.js
Download https://deno.land/std@0.115.1/async/deadline.ts
Download https://deno.land/std@0.115.1/async/debounce.ts
Download https://deno.land/std@0.115.1/async/deferred.ts
Download https://deno.land/std@0.115.1/async/delay.ts
Download https://deno.land/std@0.115.1/async/mux_async_iterator.ts
Download https://deno.land/std@0.115.1/async/pool.ts
Download https://deno.land/std@0.115.1/async/tee.ts
Download https://deno.land/std@0.115.1/http/http_status.ts
Download https://deno.land/std@0.115.1/testing/asserts.ts
Download https://deno.land/x/opine@1.9.1/src/methods.ts
Download https://deno.land/x/opine@1.9.1/src/utils/mergeDescriptors.ts
Download https://jspm.dev/superagent@6.1.0
Download https://jspm.dev/npm:component-emitter@1!cjs
Download https://jspm.dev/npm:fast-safe-stringify@2!cjs
Download https://jspm.dev/npm:qs@6!cjs
Download https://jspm.dev/npm:superagent@6.1.0!cjs
Download https://jspm.dev/npm:superagent@6.1.0/lib/agent-base!cjs
Download https://jspm.dev/npm:superagent@6.1.0/lib/is-object!cjs
Download https://jspm.dev/npm:superagent@6.1.0/lib/request-base!cjs
Download https://jspm.dev/npm:superagent@6.1.0/lib/response-base!cjs
Download https://jspm.dev/npm:superagent@6.1.0/lib/utils!cjs
Download https://jspm.dev/npm:fast-safe-stringify@2.1.1!cjs
Download https://jspm.dev/npm:component-emitter@1.3.0!cjs
Download https://jspm.dev/npm:qs@6.10.1!cjs
Download https://jspm.dev/npm:qs@6.10.1/_/9b9acba6.js
Download https://jspm.dev/npm:qs@6.10.1/lib/parse!cjs
Download https://jspm.dev/npm:qs@6.10.1/lib/stringify!cjs
Download https://jspm.dev/npm:side-channel@1!cjs
Download https://jspm.dev/npm:call-bind@1/callBound!cjs
Download https://jspm.dev/npm:get-intrinsic@1!cjs
Download https://jspm.dev/npm:object-inspect@1!cjs
Download https://jspm.dev/npm:side-channel@1.0.4!cjs
Download https://jspm.dev/npm:function-bind@1!cjs
Download https://jspm.dev/npm:get-intrinsic@1.1.1!cjs
Download https://jspm.dev/npm:has-symbols@1!cjs
Download https://jspm.dev/npm:has@1!cjs
Download https://jspm.dev/npm:call-bind@1.0.2!cjs
Download https://jspm.dev/npm:call-bind@1.0.2/callBound!cjs
Download https://jspm.dev/npm:object-inspect@1.11.0!cjs
Download https://jspm.dev/npm:has-symbols@1.0.2!cjs
Download https://jspm.dev/npm:has-symbols@1.0.2/shams!cjs
Download https://jspm.dev/npm:function-bind@1.1.1!cjs
Download https://jspm.dev/npm:has@1.0.3!cjs
Download https://deno.land/std@0.115.1/fmt/colors.ts
Download https://deno.land/std@0.115.1/testing/_diff.ts
Check file:///Users/craigmorten/git/asos-craigmorten/superoak/mod.ts
thread 'main' panicked at 'internal error: entered unreachable code: unexpected missing emit: https://deno.land/x/superdeno@4.7.0/deps.ts', cli/proc_state.rs:544:15
stack backtrace:
   0:        0x10ef5ad3f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf1ee81583766da09
   1:        0x10e8209fb - core::fmt::write::h8dc218a6eb2426a8
   2:        0x10ef5921a - std::io::Write::write_fmt::heef3e26d448f8c80
   3:        0x10ef59e95 - std::panicking::default_hook::{{closure}}::h4b7de73863cf8582
   4:        0x10ef595f7 - std::panicking::default_hook::h778da672ce56a579
   5:        0x10e7dc106 - deno::setup_exit_process_panic_hook::{{closure}}::hfe9a9204138a8142
   6:        0x10ef590c9 - std::panicking::rust_panic_with_hook::ha5146542f697c8af
   7:        0x10ef7a0ea - std::panicking::begin_panic_handler::{{closure}}::hd8bf1442c82bf304
   8:        0x10ef7a069 - std::sys_common::backtrace::__rust_end_short_backtrace::h73a3d0f5ab35b08b
   9:        0x10ef7a025 - _rust_begin_unwind
  10:        0x11013df5f - core::panicking::panic_fmt::h16f19dd104182e80
  11:        0x10e594e07 - deno::proc_state::ProcState::prepare_module_load::{{closure}}::h7ff3d5c5a5f7f040
  12:        0x10e5926be - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h95f4f81560e11b4e
  13:        0x10e5924ad - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hba28f970eba75693
  14:        0x10e4ac426 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h011315fb8be13289
  15:        0x10e4ab1a9 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1d99aa03600e7c07
  16:        0x10e5b8591 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hac80c5777a31e8a2
  17:        0x10e5b5f04 - deno::run_command::{{closure}}::h26541ce94df71d31
  18:        0x10e5b4b61 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h95b64f207b752c8f
  19:        0x10e4822e7 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hcd0e8c40af4c0e1c
  20:        0x10e473929 - deno::main::heb40db0691ea69e2
  21:        0x10e46f24c - std::sys_common::backtrace::__rust_begin_short_backtrace::h88e6039fee8e0e1b
  22:        0x10e46f4a0 - _main
make[1]: *** [build] Error 1
make: *** [ci] Error 2
```

Can confirm deleting the `DENO_DIR` appears to have resolved in this instance.

--- a/cli/emit.rs
+++ b/cli/emit.rs
@@ -304,6 +304,8 @@ pub(crate) struct CheckOptions {
   pub maybe_config_specifier: Option<ModuleSpecifier>,
   /// The derived tsconfig that should be used when checking.
   pub ts_config: TsConfig,
+  /// If true, existing `.tsbuildinfo` files will be ignored.
+  pub reload: bool,
 }
 
 /// The result of a check or emit of a module graph. Note that the actual
@@ -331,8 +333,11 @@ pub(crate) fn check_and_maybe_emit(
   // while there might be multiple roots, we can't "merge" the build info, so we
   // try to retrieve the build info for first root, which is the most common use
   // case.
-  let maybe_tsbuildinfo =
-    cache.get(CacheType::TypeScriptBuildInfo, &graph.roots[0]);
+  let maybe_tsbuildinfo = if options.reload {
+    None
+  } else {
+    cache.get(CacheType::TypeScriptBuildInfo, &graph.roots[0])
+  };
   // to make tsc build info work, we need to consistently hash modules, so that
   // tsc can better determine if an emit is still valid or not, so we provide
   // that data here.
@@ -352,18 +357,18 @@ pub(crate) fn check_and_maybe_emit(
     root_names,
   })?;
 
-  if let Some(info) = &response.maybe_tsbuildinfo {
-    // while we retrieve the build info for just the first module, it can be
-    // used for all the roots in the graph, so we will cache it for all roots
-    for root in &graph.roots {
-      cache.set(CacheType::TypeScriptBuildInfo, root, info.clone())?;
-    }
-  }
   // sometimes we want to emit when there are diagnostics, and sometimes we
   // don't. tsc will always return an emit if there are diagnostics
   if (response.diagnostics.is_empty() || options.emit_with_diagnostics)
     && !response.emitted_files.is_empty()
   {
+    if let Some(info) = &response.maybe_tsbuildinfo {
+      // while we retrieve the build info for just the first module, it can be
+      // used for all the roots in the graph, so we will cache it for all roots
+      for root in &graph.roots {
+        cache.set(CacheType::TypeScriptBuildInfo, root, info.clone())?;
+      }
+    }
     for emit in response.emitted_files.into_iter() {
       if let Some(specifiers) = emit.maybe_specifiers {
         assert!(specifiers.len() == 1);

--- a/cli/main.rs
+++ b/cli/main.rs
@@ -724,6 +724,7 @@ async fn create_graph_and_maybe_check(
         emit_with_diagnostics: false,
         maybe_config_specifier,
         ts_config,
+        reload: ps.flags.reload,
       },
     )?;
     debug!("{}", check_result.stats);

--- a/cli/ops/runtime_compiler.rs
+++ b/cli/ops/runtime_compiler.rs
@@ -252,6 +252,7 @@ async fn op_emit(
           emit_with_diagnostics: true,
           maybe_config_specifier: None,
           ts_config,
+          reload: true,
         },
       )?;
       (emit_result.diagnostics, emit_result.stats)
@@ -271,6 +272,7 @@ async fn op_emit(
           emit_with_diagnostics: true,
           maybe_config_specifier: None,
           ts_config: ts_config.clone(),
+          reload: true,
         },
       )?;
       (emit_result.diagnostics, emit_result.stats)

--- a/cli/proc_state.rs
+++ b/cli/proc_state.rs
@@ -481,6 +481,7 @@ impl ProcState {
           emit_with_diagnostics: false,
           maybe_config_specifier,
           ts_config,
+          reload: self.flags.reload,
         };
         for root in &graph.roots {
           let root_str = root.to_string();

--- a/cli/tests/integration/run_tests.rs
+++ b/cli/tests/integration/run_tests.rs
@@ -2305,3 +2305,39 @@ fn issue12740() {
     .unwrap();
   assert!(!status.success());
 }
+
+/// Regression test for https://github.com/denoland/deno/issues/12807.
+#[test]
+fn issue12807() {
+  let mod_dir = TempDir::new().expect("tempdir fail");
+  let mod1_path = mod_dir.path().join("mod1.ts");
+  let mod2_path = mod_dir.path().join("mod2.ts");
+  let mut deno_cmd = util::deno_cmd();
+  // With a fresh `DENO_DIR`, run a module with a dependency and a type error.
+  std::fs::write(&mod1_path, "import './mod2.ts'; Deno.exit('0');").unwrap();
+  std::fs::write(&mod2_path, "console.log('Hello, world!');").unwrap();
+  let status = deno_cmd
+    .current_dir(util::testdata_path())
+    .arg("run")
+    .arg(&mod1_path)
+    .stderr(std::process::Stdio::null())
+    .stdout(std::process::Stdio::null())
+    .spawn()
+    .unwrap()
+    .wait()
+    .unwrap();
+  assert!(!status.success());
+  // Fix the type error and run again.
+  std::fs::write(&mod1_path, "import './mod2.ts'; Deno.exit(0);").unwrap();
+  let status = deno_cmd
+    .current_dir(util::testdata_path())
+    .arg("run")
+    .arg(&mod1_path)
+    .stderr(std::process::Stdio::null())
+    .stdout(std::process::Stdio::null())
+    .spawn()
+    .unwrap()
+    .wait()
+    .unwrap();
+  assert!(status.success());
+}

